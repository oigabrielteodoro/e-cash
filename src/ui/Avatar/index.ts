export * from './Avatar'
export * from './NotFoundAvatar'
