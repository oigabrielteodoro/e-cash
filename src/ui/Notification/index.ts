export const NOTIFICATION_EVENT_NAME = 'notification'
export const EXPIRES_NOTIFICATION_IN_SECONDS = 1000 * 3 // 3 seconds

export * from './Notification'
export * from './NotificationContainer'
