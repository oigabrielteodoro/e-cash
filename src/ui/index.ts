export * from './Style'
export * from './Input'
export * from './Button'
export * from './PasswordInput'
export * from './Toast'
export * from './Logo'
export * from './SideBar'
export * from './AppLayout'
export * from './Tooltip'
export * from './Avatar'
export * from './ShimmerEffect'
export * from './Popover'
export * from './Grid'
