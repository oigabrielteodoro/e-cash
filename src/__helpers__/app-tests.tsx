import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
