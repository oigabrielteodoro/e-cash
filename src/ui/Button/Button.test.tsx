import React from 'react'
import { render, screen, userEvent } from '__helpers__/app-tests'

import { Button } from 'ui'

describe('Button', () => {
  it('should be able match snapshot', () => {
    const { container } = render(<Button>Click me</Button>)

    expect(container).toMatchSnapshot()
  })

  it('should be able render correctly', () => {
    render(<Button>Click me</Button>)

    expect(
      screen.getByRole('button', {
        name: /Click me/i,
      }),
    ).toBeInTheDocument()
  })

  it('should be able render correctly when is loading', async () => {
    render(<Button loading>Click me</Button>)

    expect(
      screen.queryByRole('button', {
        name: /Click me/i,
      }),
    ).not.toBeInTheDocument()

    expect(await screen.findByLabelText('loading')).toBeInTheDocument()
    expect(screen.getByRole('button')).toBeDisabled()
  })

  it('should be able called onClick when user click', async () => {
    const onClick = jest.fn()

    render(<Button onClick={onClick}>Click me</Button>)

    const button = screen.getByRole('button', {
      name: /Click me/i,
    })

    expect(button).toBeInTheDocument()

    userEvent.click(button)

    expect(onClick).toHaveBeenCalled()
  })
})
