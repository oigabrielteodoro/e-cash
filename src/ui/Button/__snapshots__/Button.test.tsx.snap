// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should be able match snapshot 1`] = `
.c0 {
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #ffffff;
  border-radius: 0.5rem;
  font-weight: 500;
  font-size: 1.6rem;
  -webkit-transition: 300ms;
  transition: 300ms;
  border: 0.1rem solid transparent;
  width: 100%;
  padding: 1.2rem 3.2rem;
  background: #3a67e1;
}

.c0:hover {
  opacity: 0.9;
  box-shadow: 0 0 0 0.4rem #a4b7ea30;
}

.c0:disabled {
  cursor: no-drop;
  color: #a1a7ae;
  background: #f5f5f5;
  border-color: #cfd7e3;
}

.c0:disabled:hover {
  box-shadow: 0 0 0 0.4rem #a1a7ae20;
}

<div>
  <button
    class="c0"
    type="button"
  >
    Click me
  </button>
  <div
    class="Toastify"
  />
</div>
`;
